sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp

```diff
@@ -0,0 +1,777 @@
+  ZeStruct<ze_fence_desc_t> ZeFenceDesc;
+  ur_command_list_ptr_t CommandListPtr;
+
+  ZE2UR_CALL(zeFenceCreate, (ZeCommandQueue, &ZeFenceDesc, &ZeFence));
+  // TODO: Refactor so requiring a map iterator is not required here, currently
+  // required for executeCommandList though.
+  ZeStruct<ze_command_queue_desc_t> ZeQueueDesc;
+  ZeQueueDesc.ordinal = QueueGroupOrdinal;
+  CommandListPtr = hCommandBuffer->CommandListMap.insert(
+      std::pair<ze_command_list_handle_t, pi_command_list_info_t>(
+          hCommandBuffer->ZeCommandList,
+          {ZeFence, false, false, ZeCommandQueue, ZeQueueDesc}));
+
+  // Previous execution will have closed the command list, we need to reopen
+  // it otherwise calling `executeCommandList` will return early.
+  CommandListPtr->second.IsClosed = false;
+  CommandListPtr->second.ZeFenceInUse = true;
+
+  // Create command-list to execute before `CommandListPtr` and will signal
+  // when `EventWaitList` dependencies are complete.
+  ur_command_list_ptr_t WaitCommandList{};
+  if (numEventsInWaitList) {
+    _ur_ze_event_list_t TmpWaitList;
+    if (auto Res = TmpWaitList.createAndRetainUrZeEventList(
+            numEventsInWaitList, phEventWaitList, hQueue, UseCopyEngine))
+      return Res;
+
+    if (auto Res = hQueue->Context->getAvailableCommandList(
+            hQueue, WaitCommandList, false, false))
+      return Res;
+
+    // Update the WaitList of the Wait Event
+    // Events are appended to the WaitList if the WaitList is not empty
+    if (hCommandBuffer->WaitEvent->WaitList.isEmpty())
+      hCommandBuffer->WaitEvent->WaitList = TmpWaitList;
+    else
+      hCommandBuffer->WaitEvent->WaitList.insert(TmpWaitList);
+
+    ZE2UR_CALL(zeCommandListAppendBarrier,
+               (WaitCommandList->first, hCommandBuffer->WaitEvent->ZeEvent,
+                hCommandBuffer->WaitEvent->WaitList.Length,
+                hCommandBuffer->WaitEvent->WaitList.ZeEventList));
+  } else {
+    if (auto Res = hQueue->Context->getAvailableCommandList(
+            hQueue, WaitCommandList, false, false))
+      return Res;
+
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
+               (WaitCommandList->first, hCommandBuffer->WaitEvent->ZeEvent));
+  }
+
+  // Execution event for this enqueue of the PI command-buffer
+  ur_event_handle_t RetEvent{};
+  // Create a command-list to signal RetEvent on completion
+  ur_command_list_ptr_t SignalCommandList{};
+  if (phEvent) {
+    if (auto Res = hQueue->Context->getAvailableCommandList(
+            hQueue, SignalCommandList, false, false))
+      return Res;
+
+    if (auto Res = createEventAndAssociateQueue(
+            hQueue, &RetEvent, UR_COMMAND_COMMAND_BUFFER_ENQUEUE_EXP,
+            SignalCommandList, false))
+      return Res;
+
+    ZE2UR_CALL(zeCommandListAppendBarrier,
+               (SignalCommandList->first, RetEvent->ZeEvent, 1,
+                &(hCommandBuffer->SignalEvent->ZeEvent)));
+  }
+
+  // Execution our command-lists asynchronously
```

Comment:
ah, so we have 3 lists, one for waiting for the dependencies for the command buffer, the actual command buffer, and another for signaling the command buffer has completed. Why cannot we do all in a single command list?

Reply:
I tried to document this at the top of the file https://github.com/intel/llvm/blob/2e3e89bd375f68da8172633df4ab91f5a1f5bf99/sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp#L12-L69

If that comment could be improved, or you'd prefer documentation a different way, please let me know. Although we do have a design document we've not upstreamed yet.

The UR interface [urcommandbufferenqueueexp](https://oneapi-src.github.io/unified-runtime/core/api.html#urcommandbufferenqueueexp) for submitting a UR command-buffer takes a list of events to wait on, and returns an event representing the completion of that particular submission of the command-buffer. 
  
However, in `zeCommandQueueExecuteCommandLists` there are no parameters to take a waitlist and also the only sync primitive returned is to block on host (which we don't want to do).  So if we used a single L0 command-list then we can't update it with the L0 wait events for that particular enqueue as it will already be closed, and the event returned would always be complete, even for subsequent submissions of the same command-list with `urcommandbufferenqueueexp`

Reply:
thanks @EwanC . Indeed, zeCommandQueueExecuteCommandLists, but we can do that in the same list. So instead of doing:

create prefix list
append wait events to prefix list
create commands list
append command list with all the commands
create suffix list
append signal events to suffix list

execute prefix list
execute commands list
execute suffix list


we could just do:

create command list
append wait events to prefix list
append command list with all the commands
append signal events to suffix list

execute commands list


this way you are only creating 1 single list and achieving same functionality. There are two main issues with the current approach:

1.
Each command list requires internal resources in the L0 driver, so for each command buffer, we are creating x3 the resources in the L0 driver, which means we could exhaust memory if we have too many graphs on the flight.

2.
Each executeCommandList call introduces a serialization in the submission pipeline that is heavier than having a barrier or a waitForEvents on the same list. Which that means is that we would have additional latency when executing graphs.

Reply:
ah, I think I got it now.

The command list holding the commands for the command buffer was in some point before closed. Now, we want to reuse it (so it is still closed). However, every time we want to resubmit it, it could have different wait events and different signal event. Since the command buffer list was already closed, we cannot append those there, hence the need for separate lists.

@EwanC is my understanding correct?

Reply:
> @EwanC is my understanding correct?

Yes, that's it exactly 🙂 Thanks for taking the time to understand the problem and our proposed solution.

> Each command list requires internal resources in the L0 driver, so for each command buffer, we are creating x3 the resources in the L0 driver, which means we could exhaust memory if we have too many graphs on the flight.

This is indeed a concern. I'd hope that SYCL users wouldn't be submitting lots of small graphs, as this diminishes the benefits of the extension grouping together individual SYCL commands calls. However, depending on the characteristics of a SYCL graph in the future we envisage that the SYCL runtime could partition a single SYCL-Graph into multiple
underlying UR command-buffers, e.g on SYCL host-tasks (which we haven't implemented yet). Once we start partitioning a graph, for every partition of the graph we'll be paying this 3x resource cost, so for a large graph with lots of host-tasks this could become an issue.

> Each executeCommandList call introduces a serialization in the submission pipeline that is heavier than having a barrier or a waitForEvents on the same list. Which that means is that we would have additional latency when executing graphs.

[zeCommandQueueExecuteCommandLists](https://spec.oneapi.io/level-zero/latest/core/api.html#zecommandqueueexecutecommandlists) can be passed a list of command-list handles, but currently `ur_queue_handle_t_::executeCommandList` only executes a single list at a time. Maybe we can provide an equivalent queue method to `ur_queue_handle_t_::executeCommandList` that instead operates on a list of command-lists to be passed to `zeCommandQueueExecuteCommandLists` to mitigate this issue. Although I don't understand the nuance of what `ur_queue_handle_t_::executeCommandList` is doing, so the maybe this wouldn't be feasible.

I've documented these drawbacks in code comments for the moment (and will put in design doc), as I'm not sure if they block this PR , or are improvements we can do as follow-up iterations.

Reply:
thanks @EwanC . passing several or one list to zeCommandQueueExecuteCommandLists has the same effect as calling multiple times zeCommandQueueExecuteCommandLists , in any case they are serialized.

Based on current support we have in L0, yes, this is the best we can do. And as you say, it could be a problem for small graphs. For now, this gives us good initial support, but it would be something that we need to think at the L0 level to see if there could be better alternatives.

Reply:
> passing several or one list to zeCommandQueueExecuteCommandLists has the same effect as calling multiple times zeCommandQueueExecuteCommandLists , in any case they are serialized.

ah, that's good to know. 

> Based on current support we have in L0, yes, this is the best we can do. And as you say, it could be a problem for small graphs. For now, this gives us good initial support, but it would be something that we need to think at the L0 level to see if there could be better alternatives.

We will get in touch with the owners of the L0 standard and discuss how we can solve this going forward.